testmodule/testErc20.nim

```diff
@@ -0,0 +1,99 @@
+
+suite "ERC20":
+
+  var token, token1: Erc20Token
+  var testToken: TestToken
+  var provider: JsonRpcProvider
+  var snapshot: JsonNode
+  var accounts: seq[Address]
+
+  setup:
+    provider = JsonRpcProvider.new("ws://localhost:8545")
+    snapshot = await provider.send("evm_snapshot")
+    accounts = await provider.listAccounts()
+    let deployment = readDeployment()
+    testToken = TestToken.new(!deployment.address(TestToken), provider.getSigner())
+    token = Erc20Token.new(!deployment.address(TestToken), provider.getSigner())
+
+    let signer1 = provider.getSigner(accounts[1])
+    token1 = Erc20Token.new(!deployment.address(TestToken), signer1)
+
+  teardown:
+    discard await provider.send("evm_revert", @[snapshot])
+
+  test "retrieves basic information":
+    check (await token.name()) == "TestToken"
+    check (await token.symbol()) == "TST"
+    check (await token.decimals()) == 12
+    check (await token.totalSupply()) == 0.u256
+    check (await token.balanceOf(accounts[0])) == 0.u256
+    check (await token.allowance(accounts[0], accounts[1])) == 0.u256
+
+  test "transfer tokens":
+    check (await token.balanceOf(accounts[0])) == 0.u256
+    check (await token.allowance(accounts[0], accounts[1])) == 0.u256
+
+    discard await testToken.mint(accounts[0], 100.u256)
+
+    check (await token.totalSupply()) == 100.u256
+    check (await token.balanceOf(accounts[0])) == 100.u256
+    check (await token.balanceOf(accounts[1])) == 0.u256
+
+    await token.transfer(accounts[1], 50.u256)
+
+    check (await token.balanceOf(accounts[0])) == 50.u256
+    check (await token.balanceOf(accounts[1])) == 50.u256
+
+  test "approve tokens":
+    check (await token.balanceOf(accounts[0])) == 0.u256
+    check (await token.allowance(accounts[0], accounts[1])) == 0.u256
+
+    discard await testToken.mint(accounts[0], 100.u256)
+
+    check (await token.totalSupply()) == 100.u256
+    check (await token.balanceOf(accounts[0])) == 100.u256
+    check (await token.balanceOf(accounts[1])) == 0.u256
+
+    await token.approve(accounts[1], 50.u256)
+
+    check (await token.allowance(accounts[0], accounts[1])) == 50.u256
+    check (await token.balanceOf(accounts[0])) == 100.u256
+    check (await token.balanceOf(accounts[1])) == 0.u256
+
+  test "transferFrom tokens":
+    check (await token.balanceOf(accounts[0])) == 0.u256
+    check (await token.allowance(accounts[0], accounts[1])) == 0.u256
+
+    discard await testToken.mint(accounts[0], 100.u256)
+
+    check (await token.totalSupply()) == 100.u256
+    check (await token.balanceOf(accounts[0])) == 100.u256
+    check (await token.balanceOf(accounts[1])) == 0.u256
+
+    await token.approve(accounts[1], 50.u256)
+
+    check (await token.allowance(accounts[0], accounts[1])) == 50.u256
+    check (await token.balanceOf(accounts[0])) == 100.u256
+    check (await token.balanceOf(accounts[1])) == 0.u256
+
+    await token1.transferFrom(accounts[0], accounts[1], 50.u256)
```

Comment:
When calling `approve` above, the current caller is `account[0]`, right? And so this is saying that `account[0]` allows `account[1]` to transfer 50 tokens (on behalf of `account[0]`), no?

In my mind, it should follow that we should switch to `account[1]` and then attempt to transfer `account[0]`'s tokens in `transferFrom`. I could be reading this completely wrong though 😅

Reply:
Yes, this would probably be clearer:

```suggestion
    let signer1 = provider.getSigner(accounts[1])
    await token.connect(signer1).transferFrom(accounts[0], accounts[1], 50.u256)
```

It would also remove the need for the `token1` variable.

Reply:
Ahh I see. I missed the `token1.transferFrom`. Yes, it would be clearer to have one token contract instance and switch signers as suggested.